Fix/doctor dashboard

V1OL3TF0X opened the pull request:
-added spinner behaviour so it behaves like a good boy
-got the Visit list to show up instead of hiding deep in the data
-fixed date formatting because it sucked (and wasn't compatible with the specification)
-changed query so it shows slots from beginning of the start day to the end of the end day, instead of the start of the end day which would be dumb

V1OL3TF0X committed:
fixes to DoctorDashboard API and look

-added spinner and await to EnterSchedulePage so it waits for all the requests to process before navigating
- fixed doctor post VaccinationSlot API
- fixed typos preventing setVisits from properly checking response status

V1OL3TF0X committed:
finally got visits to show up

I did it

V1OL3TF0X committed:
minor UI fix

set page to 1 on any other parameter change

V1OL3TF0X committed:
default start date change

-start date - today
end - week from now
-can't see slots that are earlier than today (because why would he)

adambiel22 commented on src/components/PatientVisit.tsx:
```diff
@@ -24,6 +24,7 @@ const PatientVisitField: React.FC<patientVisitProps> = (props) => {
             else
```
Odnośnie powiadomień, które są wypisywane na konsole w linijkach wyżej:
Unable to delete - time slot already reserved
Unable to delete - slot is too near

Dobrze by było gdyby one nie były wypisywane na konsoli, tylko aby użytkownik również wiedział czemu się nie powiodło usunięcie.

V1OL3TF0X replied:
Jasne, ma sens

filipspych replied:
Można chopciaz jakiś "alert" czy też równie prymitywny mechanizm (czyli zmiana paru znaków w linijce console.log -> alert)

filipspych replied:
A dobra, widzę że już poprawione nawet lepiej niż przez alert. Super 💪

IMO

adambiel22 commented on src/components/PatientVisit.tsx:
```diff
@@ -24,6 +24,7 @@ const PatientVisitField: React.FC<patientVisitProps> = (props) => {
             else
                 props.remove(props.index);
```
Tutaj też moim zdaniem dobrze by powiadomić użytkownika, że się nie udało usunąć slotu. Tylko to już w komponencie wyżej - DoctorDashboard - w tej funkcji remove

adambiel22 commented on src/pages/DoctorDashboard.tsx:
```diff
@@ -1,41 +1,58 @@
         { name: 'Reserved', value: '1' },
     ];
-
+    const min = function (a: number, b: number) {
```
Krótki research wykazał, że istnieje taka funkcja jak Math.min: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/min

adambiel22 commented on src/pages/DoctorDashboard.tsx:
```diff
@@ -1,41 +1,58 @@
     let auth: AuthContextType = useAuth();
 
-    function getVisits(): Visit[] {
+    const getVisits = useCallback(async () =>{
```
Poddaję pod wątpliwość: po co tutaj został użyty useCallback? Czy jesteś w stanie to uzasadnić?

Intuicja mi podpowiada, że useEffect spełni tę samą rolę. Zwłaszcza, że podejrzanie tak samo wygląda drugi argument obu tych funkcji - deps:[startDate, endDate, reserved, page]

adamczykt replied:
Też chciałbym zapytać o ten wybór.

V1OL3TF0X replied:
useEffect nie może dostać asynchronicznej funkcji jako argumentu, useCallback już tak
A potrzebuję async żeby poczekać na odpowiedzi

adambiel22 commented on src/pages/DoctorDashboard.tsx:
```diff
@@ -1,41 +1,58 @@
+                    }})
+                console.log(visits);
+                setVisits(visits);
+                console.log(Visits);
```
Sporo wypisywania tych wizyt na konsolę. Może by jednak je usunąć, bo rozumiem, że używałeś ich do debugowania, ale teraz chyba są średnio potrzebne.

adambiel22 commented on src/pages/DoctorDashboard.tsx:
```diff
@@ -46,19 +63,19 @@ function DoctorDashboard() {
+    }, [startDate, endDate, reserved, page])
     useEffect(() => {
-        setVisits(getVisits());
+        getVisits();
```
Tutaj intelliJ podkreśla mi i mówi: "Promise returned from getVisits is ignored". Moim zdaniem  getVisits nie musi być zwracać promisa.

adambiel22 replied:
Ogólnie myślę, że:
- getVisits może być zwykłą funkcją  - nie trzeba wywoływać useCallback), nie musi być async
- w linii 40 nie musi być await

To jest moje zdanie. Jeśli zrobiłeś tak z jakichś przyczyn to jakich?

filipspych replied:
Podpisuję się pod pytaniem Adama (zwłaszcza pod ostatnim zdaniem).

IMO

V1OL3TF0X replied:
Okej, to sprawdzę czy działa bez, bo to jest stachoverflowowe rozwiązanie na fetch w useEffect. Jak zadziała bez useCallback, to usunę

adambiel22 commented on src/pages/DoctorDashboard.tsx:
```diff
@@ -1,41 +1,58 @@
+
         setLoading(true);
-        getSlots(startDate, endDate, reserved, auth.token, page).then((response) => {
+        await getSlots(startDate, endDate, reserved, auth.token, page).then((response) => {
```
Zastanawiam się po co tutaj jest ten await? getSlots zwraca promisa i obsługuje się go w metodach then catch finally.

adambiel22 commented on src/pages/DoctorDashboard.tsx:
```diff
@@ -46,19 +63,19 @@ function DoctorDashboard() {
-            setVisits(getVisits());
+            getVisits();
         }).catch((reason => {
             console.log(reason);
```
Tutaj fajnie by jednak powiadomić użytkownika, że nie udało się usunąć

adambiel22 commented on src/pages/DoctorDashboard.tsx:
```diff
@@ -71,19 +88,25 @@ function DoctorDashboard() {
                     <Row style={{padding: '2px'}}>
                         <Col>To:</Col>
                         <Col><DatePicker
+                            disabled={loading}
```
Tutaj taka uwaga edycyjna. To się trudniej czyta:

![image](https://user-images.githubusercontent.com/51017764/163666234-9b26d66c-9bd1-4619-b074-5915d3b3a060.png)

od tego:

![image](https://user-images.githubusercontent.com/51017764/163666242-78f2f29d-5243-4f13-8a1b-7cc86d9d3424.png)

adambiel22 replied:
W IntelliJ jest skrót klawiszowy crl + alt + L, który formatuje dokument. Ale i tak czasami trzeba mu pomagać

adambiel22 commented on src/pages/DoctorDashboard.tsx:
```diff
@@ -112,15 +139,15 @@ function DoctorDashboard() {
 
             <Container>
-                <Col className={`d-flex${Visits.length === 0 ? '' : '-nowrap'} justify-content-center`}>{Visits.length === 0 ? (loading ? <Spinner animation='border'/> : <div>{error}</div>) : Visits.map((field, index) => {
+                <Col className={`d-flex${Visits.length === 0  && !loading ? '' : '-nowrap'} justify-content-center`}>{loading ?  <Spinner animation='border'/> : (Visits.length===0 ?  <div>{error}</div> : Visits.map((field, index) => {
```
Tu jest jakaś kompletna wężownia XD Nic nie da się zrozumieć.

V1OL3TF0X replied:
Bo paskuda html nie działa jak chcę
I errory środkuje na d-flex a sloty na d-flex-nowrap

V1OL3TF0X replied:
Spróbuję uczytelnić

adambiel22 commented on src/pages/DoctorDashboard.tsx:
```diff
@@ -112,15 +139,15 @@ function DoctorDashboard() {
 
             <Container>
-                <Col className={`d-flex${Visits.length === 0 ? '' : '-nowrap'} justify-content-center`}>{Visits.length === 0 ? (loading ? <Spinner animation='border'/> : <div>{error}</div>) : Visits.map((field, index) => {
+                <Col className={`d-flex${Visits.length === 0  && !loading ? '' : '-nowrap'} justify-content-center`}>{loading ?  <Spinner animation='border'/> : (Visits.length===0 ?  <div>{error}</div> : Visits.map((field, index) => {
```
<Container>
                <Col
                    className={`d-flex${Visits.length === 0 && !loading ? '' : '-nowrap'} justify-content-center`}
                >
                    {loading
                        ? <Spinner animation='border'/>
                        : (Visits.length === 0
                            ? <div>{error}</div>
                            : Visits.map((field, index) => {
                                return <PatientVisitField
                                    key={`visit_${index}`}
                                    visit={field} index={index}
                                    remove={remove}
                                />
                            }))
                    }
                </Col>
            </Container>
```suggestion
                <Col className={`d-flex${Visits.length === 0  && !loading ? '' : '-nowrap'} justify-content-center`}>{loading ?  <Spinner animation='border'/> : (Visits.length===0 ?  <div>{error}</div> : Visits.map((field, index) => {
```

adambiel22 replied:
Może jakoś tak jak wyżej?

adambiel22 commented on src/pages/DoctorDashboard.tsx:
```diff
@@ -112,15 +139,15 @@ function DoctorDashboard() {
-                    <Col className='d-flex justify-content-center'><Button disabled={page<=1} onClick={() => setPage(page - 1)}> Previous Page</Button></Col>
-                    <Col className='d-flex justify-content-center'>{page}</Col>
-                    <Col className='d-flex justify-content-center'><Button disabled={page>=maxPage} onClick={() => setPage(page + 1)}> Next Page</Button></Col>
+                    <Col className='d-flex justify-content-center'><Button disabled={loading || page<=1} onClick={() => setPage(page - 1)}> Previous Page</Button></Col>
```
Tutaj tak samo - przydałoby się porobić kilka enterów i kliknąć crl + alt + L

adambiel22 commented on src/pages/EnterSchedulePage.tsx:
```diff
@@ -10,8 +10,9 @@ import {useNavigate} from "react-router-dom";
     const navigate = useNavigate();
-    const submitForm = (formData: DoctorScheduleForm) => {
+    const submitForm = async (formData: DoctorScheduleForm) => {
         var day = new Date(formData.week.getTime());
```
tutaj intelliJ podpowiada mi aby zamienić "var" na "const". Chyba słusznie.

adambiel22 commented on src/pages/EnterSchedulePage.tsx:
```diff
@@ -10,8 +10,9 @@ import {useNavigate} from "react-router-dom";
 function EnterSchedulePage(){
```
Nazwa tego komponentu jest całkiem myląca, co nie? To nie jest jakiś artefakt przeszłości? Bo teraz to my nigdzie nie wchodzimy więc skąd "Enter.."

Może SetSchedulePage? Oraz SetScheduleForm?

V1OL3TF0X replied:
Enter jak w "wpisz", ale mogę zrobić refactor

adambiel22 commented on src/pages/EnterSchedulePage.tsx:
```diff
@@ -10,8 +10,9 @@ import {useNavigate} from "react-router-dom";
+    const [loading, setLoading] = useState(false);
     const navigate = useNavigate();
-    const submitForm = (formData: DoctorScheduleForm) => {
+    const submitForm = async (formData: DoctorScheduleForm) => {
```
DoctorScheduleForm chyba nie jest dobrą nazwą na dane z formularza, bo obiekt typu DoctorScheduleForm raczej kojarzy się z samym formularzem a nie danymi. Może DoctorScheduleData, lub coś związanego ze slotami?

adambiel22 reviewed (commented):
Dodałem swoje uwagi. Kilka z nich dotyczy formatowania kodu. Ogólnie wydaje się, że jest sporo miejsc, w których mógłbyś użyć kilku więcej enterów oraz kliknąć ctrl+alt+L (lub odpowiednik tego na maku) - ten skrót automatycznie formatuje kod (ale on nie zrobi całego formatowania za ciebie)

V1OL3TF0X committed:
implemented requested changes

filipspych commented on src/components/PatientVisit.tsx:
```diff
@@ -6,23 +6,31 @@ import moment from 'moment';
+        if(errorMsg === '')
+            setShow(false);
+        else
+        {
```
if else oba mają jedną linijkę. jeżeli if nie ma klamry to else też nie powinien mieć

IMO

filipspych commented on src/components/PatientVisit.tsx:
```diff
@@ -51,8 +59,10 @@ const PatientVisitField: React.FC<patientVisitProps> = (props) => {
                           </>
                       )
                 })}
+                <>{error === ''? '' : <><br/>{error}</>}</>
```
haha, zauważyłem już w kilku commitach, że nie dajesz spacji przed `?` w trójargumentowym operatorze warunkowym. Nie wiem czy to specjalnie ale ma to sens i bardzo mi się podoba XD mam nadzieję że specjalnie

w sensie przeczytanie tego jak pytania pomaga szybciej zrozumieć: `lubiKawę? kawa : herabta`

@V1OL3TF0X czy to specjalnie specjalnie? fajnie : też_fajnie

IMO

filipspych commented on src/components/PatientVisit.tsx:
```diff
@@ -51,8 +59,10 @@ const PatientVisitField: React.FC<patientVisitProps> = (props) => {
+                <>{error === ''? '' : <><br/>{error}</>}</>
             </Modal.Body>
             <Modal.Footer>
+
```
dziwny enter

IMO

filipspych commented on src/logic/doctorAPI.tsx:
```diff
@@ -1,16 +1,19 @@
-    let eDate = end == null?'':`&endDate=${encodeURIComponent(end.toDateString())}`;
-    let reserved = onlyReserved === '-1'?'':`onlyReserved=${encodeURIComponent(onlyReserved)}&`
+export function getSlots(start: Date | null, end: Date | null, onlyReserved: string, authToken: string | null, page: number) {
+    if (start != null) start.setHours(0, 0, 0, 0);
```
kurcze, bardzo źle się to czyta 🙁. To (czyli linijki 6-10) powinno być wywalone do jakieś funkcji o nazwie loadDefaults albo normalize.

IMO

filipspych commented on src/logic/doctorAPI.tsx:
```diff
@@ -1,16 +1,19 @@
+    let sDate = start == null ? '' : `&startDate=${encodeURIComponent(moment(start).format("YYYY-MM-DDThh:mm:ssZ"))}`;
+    let eDate = end == null ? '' : `&endDate=${encodeURIComponent(moment(end).format("YYYY-MM-DDThh:mm:ssZ"))}`;
+    let reserved = onlyReserved === '-1' ? '' : `onlyReserved=${encodeURIComponent(onlyReserved)}&`
     return fetch(`${BASE_URL}/doctor/vaccination-slots?${reserved}page=${page}${sDate}${eDate}`,
```
Podobnie (chociaż co do tego jestem mniej pewien) samo fetch moim zdaniem powinien być w osobnej funkcji, tak żeby jedyne co jest w tej funkcji getSlots to pokazanie, że najpierw normalizujemy dane a potem fetchujemy co trzeba. Trudno jest napisać "zbyt dużo funkcji".

IMO

filipspych commented on src/pages/EnterSchedulePage.tsx:
```diff
@@ -53,32 +52,32 @@ function EnterSchedulePage(){
+                    </>
+                )
+            })} </div> : <></>}
+            {loading ? <Spinner animation="border"/> : <ScheduleForm onSubmit={submitForm}/>}
```
fajne

filipspych commented on src/logic/login.ts:
```diff
@@ -13,10 +13,10 @@ export async function login(role: Role, email: string,
-            throw response;
-        });
+        if (response.ok) return response.json()
+        if (response.status === StatusCodes.UNAUTHORIZED) {
```
skopane wcięcia, nie da się czytać ;/ te dwa ify wyglądają jakby były na równym poziomie logicznym co .then(response => {...}) a są **wewnątrz** lambdy.

plis każde IDE ma funkcje auto-formatowania przy zapisie pliku. takie małe rzeczy naprawdę spowalniają czytanie 

IMO

filipspych commented on src/pages/EnterSchedulePage.tsx:
```diff
@@ -22,29 +23,27 @@ function EnterSchedulePage(){
             ...convertSlots(new Date(day.getTime() + 5 * 86400000), formData.satSlots),
```
Wiem, że to nie ten PR, ale jak mogliśmy do tego dopuścić? (mowa oczywiście o osiemdziesiąt sześć milionów czterysta tysięcy
.
.
razy 5.) 

IMO

V1OL3TF0X replied:
Ehhhh, byłem absolutnie pewien że poprawiłem, mam od tego funkcję

filipspych reviewed (commented):
Ogólnie git, skoro działa, frontend nie wymaga jakiegoś przywiązania do bezpieczeństwa kodu.

Jednak nie mogę nie zauważyć, że jest bardzo, bardzo chaotyczne, moim zdaniem potrzeba (/chociaż/) więcej funkcji. Nazwy funkcji pełnią rolę etykietek na fragmenty kodu, które bardzo pomagają go sprawnie zrozumieć. Takie funkcje, które jedyne co robią to wywołują inne funkcje (tzw. **laundry list functions**) **NIE są złe**. 
Wprost przeciwnie, mogą dodać jakąkolwiek strukturę w kodzi, który w którym nie ma struktury narzuconej przez technologię. React, w przeciwności do naszych backendowych technologii, praktycznie nie narzuca struktury.

filipspych commented:
Mam jeszcze ogólną uwagę: przy frontendzie dobrze jest w opisie PR umieścić screenshota jak ma wyglądać to co zostało dodane, żeby po uruchomieniu wiadomo było na co patrzeć. @V1OL3TF0X @adambiel22 @adamczykt

V1OL3TF0X committed:
yet another implementation of requested changes

- changed name from "EnterSchedulePage" to "SetSchedulePage"
- vaccine may be undefined for reserved slot - accounted for
- formatting
- deleted useCallback as it's not needed

V1OL3TF0X committed:
updated components

-changed way of checking if vaccinationSlot is reserved
- updated checking for whether vaccine is set (just to be 100% sure)

V1OL3TF0X committed:
fixes

- fixed date formatting (from 12h to 24h)
- fixed routing
